### ISSUE TITLE
Feature Request: make map, mapValues, filter, ... like functions become null-safe like lodash?

### ISSUE THREAD

**Issue by ericfong** [remeda/remeda#147]
I really like "data-first", "data-last" and "fixed number of arguments" approach.

It would be better and easier to use if map, mapValues, filter, ... like functions become null-safe.

So, we can

```js
R.map(arrOrNull, ...)
```

instead of 

```js
R.map(arrOrNull || [], ...)
```

**Comment by TkDodo** [comment]
I think that's an okay change that would also be backwards compatible, provided that type inference will still work in data-last versions and pipes. If you want, you can play around with it and create a PR, maybe for one of the functions as proof of concept and we can take it from there?

**Comment by Bessonov** [comment]
I'm not sure that I like the idea. I have following reasons:
- The semantic isn't clear. Is `null` converted to and handled as an empty array?
- What's about `undefined`? Same handling as for `null`?
- It contradicts with "normal" array semantic. `null.map` is invalid and `null?.map` doesn't execute map.
- How this is handled in other libraries?

Overall, I think it will introduce silent failing errors, which are hard to debug.

**Comment by TkDodo** [comment]
All valid points. I know lodash works with falsy input. There are also some methods where the semantics are not so obvious. Like, what should `R.clone(null)` do - fallback to empty object and then clone that?

**Comment by Bessonov** [comment]
> I know lodash works with falsy input.

... and mutates passed objects. Two of the reasons why lodash shouldn't be used :)

> Like, what should R.clone(null) do - fallback to empty object and then clone that?

In this concrete example, if null is allowed at all, I would expect to get `null`. For me it's the same as `{prop: null}`. Or do I miss something :mag: ?

**Comment by TkDodo** [comment]
> Two of the reasons why lodash shouldn't be used :)

👍 

> Or do I miss something 🔍 ?

It was just my attempt at signaling why accepting falsy input for all functions might be intransparent in some cases :)

**Comment by ericfong** [comment]
Hi,

BTW, I think lodash treat `null` and `undefined` as empty array in map/filter.

I think one of the Remeda Design Goals is 'The usage must be programmer friendly, and that's more important than following XYZ paradigm strictly.' 
And my personal reason for using lodash or remeda like util instead of Array.map() just because they wrap native js func up and handle this and that (auto null handling) for me. 

I would like to know 'accepting falsy input' will it break "Manual annotation should never be required, and proper typings should infer everything"?

I am thinking to wrap up rememda functions in my own lib and do 'accepting falsy input' there. But don't know how to make type definition right. Or I should fork the library is easier?

**Comment by TkDodo** [comment]
I think you could fork the lib and try to make functions like `map` and `filter` accept `null` and `undefined` as well, then create a draft PR. Then, we can see if this concept breaks type inference, for example:

```
function myFunction(input: Array<number> | null) {
    R.pipe(
        input,
        R.map((x) => x * 2) // here, `x` should still be inferred to `number`
     )
}
```

and then we can see on which functions it would make sense.

> It contradicts with "normal" array semantic. null.map is invalid and null?.map doesn't execute map.

Regarding semantics, I would expect the above example to be roughly equivalent to:

```
input?.map(x => x * 2) ?? []
```

I think falling back to empty array is needed so that we can chain functions together, unless _all_ functions accept null / undefined as input. Unless we _want_ the null / undefined input also as output....

**Comment by andrewdhazlett** [comment]
I'd suggest having a null-safe version available as a property available on each method, a la `indexed`

**Comment by TkDodo** [comment]
> I'd suggest having a null-safe version available as a property available on each method, a la indexed

@eranhirsch what do you think?

**Comment by eranhirsch** [comment]
I think that we shouldn't. I think the comparison to lodash is misleading, as it was designed without typing in mind, so without accepting null values code could explode in runtime/production; i'm guessing that accepting null values was a safety feature.

A strongly typed library on the other hand protects you by not allowing you to make those mistakes during compile time. If a non-nullable array suddenly became nullable because of change upstream, accepting nulls like this would hide that change, and it would only be discovered in runtime/production following an observable bug. By not accepting these values you get a compile error the moment you do the change, and you can make a deliberate decision how to act to the null case.

TL;DR - Accepting null values would weaken the typing assurances that Remeda offers it's users in flavor of slightly terser code (basically it removes `?? []`) and I don't think that's a good tradeoff.

**Comment by scamden** [comment]
> A strongly typed library on the other hand protects you by not allowing you to make those mistakes during compile time.

@eranhirsch 

I agree in normal scenarios but I'm migrating a legacy code base that has both lodash receiving undefined in this call positions _and_ prodigious use of any. This leaves me in a lurch because while the types would normally be a better place to catch these. I can't trust the types so I have runtime errors in many positions where I swapped remeda for lodash on things like `filter` etc. 

Any suggestions? My only thought currently is to add `?? []` at every call site (not just the officially nullable ones) which is pretty heinous.. 

(Love the strong types btw, just in a tricky legacy situation here.)

**Comment by cjquines** [comment]
that's a good question. you could wrap all the remeda calls with a function that _is_ nullsafe, and ratchet swapping all of them out, maybe?

**Comment by eranhirsch** [comment]
I highly recommend first fixing the underlying issues with having `any` to begin with:
1. Gradually upping the strictness of typescript (e.g. https://github.com/tsconfig/bases/blob/main/bases/strictest.json)
2. Installing eslint and fixing anything from the the `recommended` config
3. Installing typescript-eslint and fixing anything from the `recommended` config
4. Adding type-checking to typescript-eslint and using `recommendedWithTypeChecking`

And only then installing remeda to replace Lodash. The gap is just too big between Lodash assumptions and Remeda assumptions. Remeda is Type-Script first, and relies heavily on the project having sound types. Said differently, from the typing perspective, assume that Remeda is a subset of Lodash, so that if your project has no typescript errors with Lodash, it doesn't mean it wouldn't have typescript errors with Remeda.

**Comment by scamden** [comment]
@eranhirsch No disagreement here and I'm already in the process. But it's a long process.

> And only then installing remeda to replace Lodash. The gap is just too big between Lodash assumptions and Remeda assumptions.

I was actually able to inventory all remeda uses and provide non-null fallbacks. Bit easier than I thought.

But I'm curious, are there other runtime differences I should be aware of?

My impression from the lodash migrate page was that Remeda is basically Lodash with more sound types. Unfortunately, it seems to actually be lodash with more sound types and _less sound_  runtime behavior, which seems worth calling out for newcomers

**Comment by eranhirsch** [comment]
> are there other runtime differences I should be aware of?

I can't think of anything specific, but it's important to note that Remeda isn't a fork of Lodash, and everything is implemented from scratch, so there might always be differences. Most utilities are simple and straightforward, so there are only a few places where they can differ, but the problem with Lodash is that there are a lot of implicit assumptions when handling edge cases (a lot of implicit type coalescing). Most of these aren't even documented or tested. TL;DR - There is always a risk when migrating to a different library.

> it seems to actually be lodash with more sound types and less sound runtime behavior.

What do you mean by less sound runtime behavior? Assuming input types are correct where do you feel the runtime can fail unexpectedly? 

> seems worth calling out for newcomers

I don't have experience using Lodash and inherited the mapping data for Lodash from other contributors. I'd love it if people could contribute more information on tackling a real-life project's migration to Remeda from Lodash. We have a whole infra on our docs page, which allows us to add "articles" to the docs page, and I would gladly accept any contributions to that section.

**Comment by scamden** [comment]
> TL;DR - There is always a risk when migrating to a different library.

I'm happy with that risk. But you made it sound like Remeda wouldn't be safe without first getting the types perfect. AFAICT this nullability thing is the only major difference at runtime with the set of utilities we are using at least. 

FWIW our tests are passing and I'm already enjoying the better types! (entries() as tuples and fromEntries() respecting tuples were two I had previously hand written, so nice to see 😄 )

> which allows us to add "articles" to the docs page, and I would gladly accept any contributions to that section.

Process was reasonably straight forward! I would be happy to write something quick if you point me to where. (Our use of lodash was luckily not especially advanced TBH though so will differ for more power users probably.)

**Comment by scamden** [comment]
oh and 

> Assuming input types are correct where do you feel the runtime can fail unexpectedly?

I meant assuming the types are possibly incorrect the runtime behavior doesn't guard as much as lodash like you mentioned above a bit. 

>  so without accepting null values code could explode in runtime/production; i'm guessing that accepting null values was a safety feature. 

It's an option to do both things: make the types especially safe _and_ guard against runtime crashes

**Comment by eranhirsch** [comment]
One of the differentiators between Lodash and Remeda is that you don't need your utilities to do checks like:

```ts
if (array === undefined) {
  return [];
}
```

When you can make sure array is never `undefined` via typing, so you end up shipping less code. This impacts both performance and bundle size. It's an intentional design decision not to protect in runtime what we can protect in compile time (aka in type-checking time).

The TL;DR is that by not accepting rogue values we provide 3 benefits over Lodash:
1. Smaller bundle
2. Less executed lines of code
3. Errors are detected during the build (while you write the code), and not during testing or production.
